photon-ml/src/main/scala/com/linkedin/photon/ml/data/KeyValueScore.scala

```diff
@@ -63,7 +63,12 @@ protected[ml] class KeyValueScore(val scores: RDD[(Long, Double)]) extends RDDLi
     val addedScores =
       this.scores
         .fullOuterJoin(that.scores)
-        .mapValues { case (thisScore, thatScore) => thisScore.getOrElse(0.0) + thatScore.getOrElse(0.0) }
+        .mapValues { case (thisScore, thatScore) =>
+          require(!thisScore.isEmpty || !thatScore.isEmpty, s"Should not have reached this state. A full outer join " +
+            s"should yield a result that has at least one of $thisScore or $thatScore non-empty")
+          thisScore.getOrElse(thatScore.get.getZeroScoreDatum()) +
+            thatScore.getOrElse(thisScore.get.getZeroScoreDatum())
```

Comment:
I did a small offline benchmark, and a direct `cogroup -> flatMapValues` was about 6 times faster than `fullOuterJoin -> mapValues`.